src/scope.rs

```diff
@@ -539,1771 +1224,877 @@ impl<P> TryCatch<'_, P> {
-            context: Default::default(),
-            escape_slot: Default::default(),
-            try_catch: Default::default(),
-            scope_type_specific_data: Default::default(),
-          },
-        );
-        Box::from_raw(self_ptr)
-      }
-    }
-  }
+#[macro_export]
+macro_rules! make_allow_javascript_execution_scope {
+  ($scope: ident, $param: expr) => {
+    let mut $scope = $crate::AllowJavascriptExecutionScope::new($param);
+    let mut $scope = {
+      let scope_pinned = unsafe { std::pin::Pin::new_unchecked(&mut $scope) };
+      scope_pinned.init()
+    };
+    let $scope = &mut $scope;
+  };
+  (let $scope: ident, $param: expr) => {
+    $crate::make_allow_javascript_execution_scope!($scope, $param);
+  };
+}
 
-  #[derive(Debug, Clone, Copy, Eq, PartialEq)]
-  enum ScopeStatus {
-    Free,
-    Current { zombie: bool },
-    Shadowed { zombie: bool },
-  }
+#[allow(unused_imports)]
+pub(crate) use make_allow_javascript_execution_scope;
 
-  impl Default for ScopeStatus {
-    fn default() -> Self {
-      Self::Free
-    }
+#[macro_export]
+macro_rules! make_escapable_handle_scope {
+  ($scope: ident, $param: expr) => {
+    let mut $scope = $crate::EscapableHandleScope::new($param);
+    let mut $scope = {
+      let scope_pinned = unsafe { std::pin::Pin::new_unchecked(&mut $scope) };
+      scope_pinned.init()
+    };
+    let $scope = &mut $scope;
+  };
+  (let $scope: ident, $param: expr) => {
+    $crate::make_escapable_handle_scope!($scope, $param);
+  };
+}
+
+#[allow(unused_imports)]
+pub(crate) use make_escapable_handle_scope;
+
+#[repr(transparent)]
+pub struct PinnedRef<'p, T>(Pin<&'p mut T>);
+
+impl<'p, T> From<Pin<&'p mut T>> for PinnedRef<'p, T> {
+  fn from(value: Pin<&'p mut T>) -> Self {
+    PinnedRef(value)
   }
+}
 
-  #[derive(Debug)]
-  enum ScopeTypeSpecificData {
-    None,
-    ContextScope {
-      _raw_context_scope: raw::ContextScope,
-    },
-    HandleScope {
-      allow_zombie: bool,
-      raw_handle_scope: raw::HandleScope,
-      raw_context_scope: Option<raw::ContextScope>,
-    },
-    EscapableHandleScope {
-      raw_handle_scope: raw::HandleScope,
-      raw_escape_slot: Option<raw::EscapeSlot>,
-    },
-    TryCatch {
-      raw_try_catch: raw::TryCatch,
-    },
-    DisallowJavascriptExecutionScope {
-      raw_scope: raw::DisallowJavascriptExecutionScope,
-    },
-    AllowJavascriptExecutionScope {
-      raw_scope: raw::AllowJavascriptExecutionScope,
-    },
+impl<T> PinnedRef<'_, T> {
+  pub fn as_mut_ref(&mut self) -> PinnedRef<'_, T> {
+    PinnedRef(self.0.as_mut())
   }
+}
 
-  impl Default for ScopeTypeSpecificData {
-    fn default() -> Self {
-      Self::None
+impl<'p, 'i> Deref for PinnedRef<'p, HandleScope<'i>> {
+  type Target = PinnedRef<'p, HandleScope<'i, ()>>;
+  fn deref(&self) -> &Self::Target {
+    unsafe {
+      std::mem::transmute::<
+        &PinnedRef<'p, HandleScope<'i>>,
+        &PinnedRef<'p, HandleScope<'i, ()>>,
+      >(self)
```

Comment:
There are many of these transmutes in this file. Maybe there can be a inline helper to reduce surface area?

```rust
#[inline(always)]
unsafe fn cast_pinned_ref<U, V>(u: &PinnedRef<'_, U>) -> &PinnedRef<'_, V> {}
#[inline(always)]
unsafe fn cast_pinned_ref_mut<U, V>(u: &mut PinnedRef<'_, U>) -> &mut PinnedRef<'_, V> {}
```